From d59cf72971f9b7110d9c179dc8ae8b7a11dbd6d2 Mon Sep 17 00:00:00 2001
From: Francis Laniel <flaniel@linux.microsoft.com>
Date: Mon, 9 Feb 2026 19:08:12 +0200
Subject: [PATCH] pkg: columns: Escape strings before printing them.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The columns output did not run any sanitization of the output before printing it
which could lead to terminal interpreting escape sequence.
Consider the following C code:
open("/etc/shadow", O_RDONLY);
// ...
open("/etc\x1b[1A/bashrc\x1b[1B\x1b[13C", O_CREAT | O_WRONLY | O_TRUNC, 0644);
The first open() opens /etc/shadow, while the second open() creates a file which
names contains shell escape sequence.
With the current behavior, the trace_open gadget output is modified by the shell
escape sequence to hide the malicious read to /etc/shadow:
RUNTIME.CONTAINERN… COMM               PID        TID TID         FD FNAME                    FPATH                    ERROR   MODE
...
poc-escape-inject   escape_inj…      25641      25641 25591        3 /etc/bashrc                                               ----------
poc-escape-inject   escape_inj…      25641      25641 25591        0 /etc                                              ENOENT  -rw-r--r--

By using strconv.Quote() on the strings which are about to be outputted, we
prevent the escape sequence to be interpreted and the output is now correct:
RUNTIME.CONTAINERN… COMM               PID        TID TID         FD FNAME                    FPATH                    ERROR   MODE
...
poc-escape-inject   escape_inj…     116656     116656 116605       3 /etc/shadow                                               ----------
poc-escape-inject   escape_inj…     116656     116656 116605       0 /etc\x1b[1A/bashrc\x1b[…                          ENOENT  -rw-r--r--

Fixes: GHSA-34r5-6j7w-235f ("Unsanitized ANSI Escape Sequences In `columns` Output Mode")
Reported-by: Nicolò Daprelà <nicolo.daprela@shielder.it>
Reported-by: Pietro Tirenna <pietro.tirenna@shielder.it>
Signed-off-by: Francis Laniel <flaniel@linux.microsoft.com>
---
 pkg/columns/formatter/textcolumns/output.go | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/pkg/columns/formatter/textcolumns/output.go b/pkg/columns/formatter/textcolumns/output.go
index 727630949fd..251c4fa20d5 100644
--- a/pkg/columns/formatter/textcolumns/output.go
+++ b/pkg/columns/formatter/textcolumns/output.go
@@ -17,6 +17,7 @@ package textcolumns
 import (
 	"bytes"
 	"io"
+	"strconv"
 	"strings"
 
 	"github.com/inspektor-gadget/inspektor-gadget/pkg/columns"
@@ -36,6 +37,11 @@ func (tf *TextColumnsFormatter[T]) buildFixedString(s string, length int, ellips
 		return ""
 	}
 
+	// Escape the string to avoid the terminal to interpret escape sequence when
+	// printing and remove first and last character as they are quotes added by
+	// strconv.Quote().
+	s = strconv.Quote(s)
+	s = s[1 : len(s)-1]
 	if !tf.options.ShouldTruncate {
 		return s
 	}
